### PR TITLE
Sanitize PR review text

### PR DESCRIPTION
**Pull request by glenn-jocher**
## Summary

Prevent internal citation artifacts and non-printing Unicode characters from appearing in AI-generated PR reviews.

## Changes

- remove orphaned web-search references such as `turn10search1` followed by a private-use terminator
- strip non-printing/control/private-use Unicode while preserving normal Unicode and Markdown whitespace
- add regression coverage and bump the package version to `0.3.17`

## Validation

- `uv run pytest tests -q` (167 passed)
- repo Ruff lint and format checks

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary

AI-generated PR review text is now sanitized before posting to remove leaked citation artifacts and unsafe Unicode characters, while preserving normal Unicode and Markdown whitespace.

### 📊 Key Changes

- Expanded `sanitize_ai_text()` to remove private-use, control, formatting, surrogate, and noncharacter Unicode values.
- Removed orphaned search references such as `turn10search1` and `turn7fetch3` when followed by private-use terminators.
- Applied sanitization to review summaries, comment messages, and comment suggestions.
- Preserved standard Unicode, newlines, tabs, zero-width joiners/non-joiners, and supported emoji flag sequences.
- Added regression coverage and bumped the package version to `0.3.17`.

### 🎯 Purpose & Impact

- Posted GitHub reviews no longer include the targeted internal citation markers or non-printing Unicode artifacts in summaries, findings, or suggestions.
- Existing readable Unicode and Markdown-relevant whitespace remain intact.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting a `ultralytics/actions` 🚀 PR! This automated message confirms your contribution was received, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/actions/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/actions` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Reviewed the version bump, citation regex, Unicode filtering, and review-output integration. The sanitizer correctly covers summaries and messages, but suggestions bypass it and can still expose the artifacts this PR intends to remove. One medium-severity integration issue remains.

<details><summary>💬 Posted 1 inline comment</summary>

- 💡 **MEDIUM** `actions/utils/openai_utils.py:80` The review path applies `sanitize_ai_text` only to `summary` and each comment's `message` (`actions/review_pr.py:703-706`); `suggestion` is still interpolated into the posted GitHub suggestion block unchanged. An AI response containing `turn10search1\ue201` or other private-use/control characters in `suggestion` will therefore still publish the artifact. Sanitize `c["suggestion"]` in the same response-normalization loop before suggestions are posted.
</details>

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 2

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Since the previous review, the suggestion-field sanitization issue is addressed and covered by regression tests. The current diff still has two edge-case regressions: supplementary-plane private-use citation terminators can leave citation tokens behind, and the broad `isprintable()` filter can corrupt composed Unicode such as ZWJ emoji. These should be fixed before merge.

<details><summary>💬 Posted 2 inline comments</summary>

- 💡 **MEDIUM** `actions/utils/openai_utils.py:74` This citation alternative only recognizes BMP private-use terminators. For example, `sanitize_ai_text("turn10search1\U000F0000")` leaves `turn10search1` behind because the generic filtering removes only the supplementary private-use character. The test suite itself uses `\U000F0000` as a private-use character, so expand the citation terminator pattern to cover the intended supplementary private-use ranges or match the token before generic private-use stripping.
- 💡 **MEDIUM** `actions/utils/openai_utils.py:80` Using `str.isprintable()` as the Unicode filter removes semantic format characters such as the zero-width joiner. For example, `sanitize_ai_text("👩‍💻")` becomes `"👩💻"`, changing normal Unicode/emoji rendering even though the PR intends to preserve readable Unicode. Filter the unwanted control/invisible characters explicitly while retaining joiners and other characters required for composed Unicode sequences.
</details>

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 3

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Since review 2, the supplementary-plane citation terminator handling, ZWJ preservation, and suggestion sanitization are addressed and covered by tests. Two remaining Unicode-sanitization issues are identified: runtime-dependent removal of valid newer characters and preservation of bidi spoofing controls.

<details><summary>💬 Posted 2 inline comments</summary>

- 💡 **MEDIUM** `actions/utils/openai_utils.py:78` Filtering category `Cn` removes valid, newly assigned Unicode characters when this package runs on an older supported Python runtime. Python 3.8 uses Unicode data version 12.1.0, so a character such as U+1FA77 (pink heart, added in Emoji 15.0) is classified as unassigned and silently removed, contradicting the stated preservation of normal Unicode. Avoid treating all `Cn` characters as unsafe, or use a version-independent denylist. ([docs.python.org](https://docs.python.org/es/3.8/library/unico…
- 💡 **MEDIUM** `actions/utils/openai_utils.py:79` The sanitizer still preserves most invisible Unicode format controls because only ZWSP and BOM are listed here. For example, U+202E RIGHT-TO-LEFT OVERRIDE and the bidi isolate controls remain in posted review text and can reorder or spoof the visible content of a finding, despite the PR's goal of removing non-printing characters. Expand the unsafe-format handling to cover these controls while explicitly retaining required composition characters such as ZWJ/ZWNJ. ([unicode.org](https://unicode.o…
</details>

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 4

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Since review 3, suggestion sanitization, supplementary-plane citation terminators, ZWJ preservation, and the previously identified bidi controls are addressed. One remaining edge case is that other invisible `Cf`/default-ignorable characters can still appear in published review text; no other issues were found.

<details><summary>💬 Posted 1 inline comment</summary>

- 💡 **MEDIUM** `actions/utils/openai_utils.py:78` The sanitizer still lets many non-printing `Cf` characters through: this set excludes `Cf`, while `_UNSAFE_FORMAT_PATTERN` only covers a subset. For example, U+FFF9–U+FFFB are `Cf` interlinear-annotation controls and U+E0001 is a `Cf` tag character, but neither is matched here, so model output containing them is posted unchanged. turn0search15 Extend the unsafe-format handling to cover the remaining invisible/default-ignorable ranges, retaining only composition controls such as ZWJ/ZWNJ where…
</details>

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 5

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Since review 4, the prior citation, suggestion, supplementary private-use, ZWJ, and broader format-control issues are addressed. One remaining gap is the unconditional preservation of the Unicode tag-character range, which allows invisible characters to remain in posted review text.

<details><summary>💬 Posted 1 inline comment</summary>

- 💡 **MEDIUM** `actions/utils/openai_utils.py:79` `_SAFE_FORMAT_CHARACTERS` unconditionally whitelists the entire U+E0020–U+E007F tag range. Consequently, an isolated invisible tag character in model output survives `sanitize_ai_text` even though the sanitizer is intended to remove non-printing characters; the current tests cover U+E0001 but not this whitelisted range. Avoid preserving the full range for arbitrary text—strip it or preserve only validated tag sequences used by composed emoji.
</details>

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 6

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Since review 5, the tag-character leak is addressed: `Cf` filtering now removes the previously whitelisted U+E0020–U+E007F range, while ZWJ/ZWNJ and Markdown newlines/tabs remain preserved. Suggestion sanitization and supplementary-plane citation handling are also covered. One remaining edge case is that Unicode noncharacters classified as `Cn` can still reach posted reviews; otherwise LGTM.

<details><summary>💬 Posted 1 inline comment</summary>

- 💡 **MEDIUM** `actions/utils/openai_utils.py:78` Because `Cn` is omitted from `_UNSAFE_UNICODE_CATEGORIES`, Unicode noncharacters such as U+FFFF are retained: `sanitize_ai_text("Finding.\uFFFF")` still contains that code point. Noncharacters share the `Cn` category with unassigned code points, so avoid filtering all `Cn` (which would remove valid newer characters) but add an explicit noncharacter-range check to prevent these non-printing values from being posted. ([unicode.org](https://www.unicode.org/faq/private_use.html?utm_source=openai))
</details>

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 7

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Since review 6, the sanitizer now handles supplementary private-use terminators, format controls, tag characters, and BMP/plane-end noncharacters. One remaining gap is that supplementary-plane code points in the U+FDD0–U+FDEF offset range can still reach posted reviews; no other issues were found.

<details><summary>💬 Posted 1 inline comment</summary>

- 💡 **MEDIUM** `actions/utils/openai_utils.py:92` The noncharacter check only applies the U+FDD0–U+FDEF range to the full code point, so supplementary-plane values such as U+1FDD0 pass through unchanged. Unicode reserves this range in the BMP only? Actually, the remaining noncharacters are the final two code points of each plane; to cover the analogous supplementary range consistently, compare the low 16 bits for the U+FDD0–U+FDEF test as well.
</details>

File: actions/utils/openai_utils.py [hunk 1]
```diff
@@ -69,14 +70,31 @@
...
+    """Strip citation tokens and unsafe Unicode characters from AI output."""
+    return (
+        "".join(
+            c
+            for c in _CITATION_PATTERN.sub("", s)
+            if c in "\n\t"
+            or c in _SAFE_FORMAT_CHARACTERS
+            or (
+                unicodedata.category(c) not in _UNSAFE_UNICODE_CATEGORIES
+                and not (0xFDD0 <= ord(c) <= 0xFDEF or (ord(c) & 0xFFFF) in (0xFFFE, 0xFFFF))
```
Comment (glenn-jocher) [inline actions/utils/openai_utils.py]: This suggested change is incorrect. The Unicode Standard defines exactly 66 noncharacters: U+FDD0–U+FDEF in the BMP only, plus U+nFFFE and U+nFFFF at the end of each plane. U+1FDD0 is not a noncharacter, so comparing the FDD0–FDEF range against the low 16 bits would remove valid/reserved supplementary code points beyond the standard noncharacter set. The current check matches the authoritative definition: https://www.unicode.org/faq/private_use.html#nonchar4

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 8

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Since review 7, the prior noncharacter concern is resolved and the earlier suggestion-field and citation-sanitization issues remain addressed. One newly identified regression remains: stripping the entire tag-character range corrupts valid subdivision-flag emoji sequences; otherwise the diff looks correct.

<details><summary>💬 Posted 1 inline comment</summary>

- 💡 **MEDIUM** `actions/utils/openai_utils.py:78` Filtering every `Cf` character except ZWJ/ZWNJ also strips the tag characters that form valid subdivision-flag emoji. For example, the Unicode England flag sequence `🏴\U000E0067\U000E0062\U000E0065\U000E006E\U000E0067\U000E007F` is reduced to `🏴`, silently corrupting normal Unicode text. Preserve only validated emoji tag sequences while still removing isolated tag characters to address the earlier leak. ([unicode.org](https://www.unicode.org/faq/languagetagging.html?utm_source=openai))
</details>

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review 9

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Since review 8, validated that the sanitizer now preserves the three standardized subdivision-flag emoji sequences while still removing isolated tag characters; prior citation, noncharacter, format-control, and suggestion-sanitization issues remain addressed. No concrete regressions found in the current diff. LGTM.

**Comment by UltralyticsAssistant** [comment]
🎉 Thanks, @glenn-jocher, for merging [PR #890](https://github.com/ultralytics/actions/pull/890)!

> “The details are not the details. They make the design.” — Charles Eames

This PR strengthens AI-generated review output by removing citation artifacts and unsafe Unicode while preserving readable Unicode, Markdown whitespace, and supported emoji sequences. The added regression coverage and `0.3.17` release help keep summaries, findings, and suggestions clean and reliable.
